packages/table/components/Table.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from "react";
+import { css } from "emotion";
+import { AutoSizer, MultiGrid } from "react-virtualized";
+import { IColumnProps, Column } from "./Column";
+import memoizeOne from "memoize-one";
+import { coreColors as Colors } from "../../shared/styles/color";
+
+export interface ITableProps {
+  fixedColumnCount: number;
+  enableFixedColumnScroll: boolean;
+  enableFixedRowScroll: boolean;
+  data: any;
+  children:
+    | Array<React.ReactElement<IColumnProps>>
+    | React.ReactElement<IColumnProps>;
+}
+
+export class Table extends React.PureComponent<ITableProps, {}> {
+  public multiGridRef: { recomputeGridSize?: any } = {};
+
+  public getData = memoizeOne(
+    (data: any[]): any[] => {
+      return [{}, ...data];
+    }
+  );
+
+  public getColumnSizes = memoizeOne(
+    (
+      children: Array<React.ReactElement<IColumnProps>>,
+      width: number
+    ): number[] => {
+      const totalColumns: number = children.length;
+      let remainingWidth: number = width;
+      return children.map((child, currentIndex) => {
+        const calculatedWidth = child.props.width({
+          width,
+          totalColumns,
+          currentIndex,
+          remainingWidth
+        });
+        remainingWidth -= calculatedWidth;
+        return calculatedWidth;
+      });
+    }
+  );
+
+  constructor(props) {
+    super(props);
+
+    this.getCell = this.getCell.bind(this);
+    this.setRef = this.setRef.bind(this);
+    this.updateGrid = this.updateGrid.bind(this);
+  }
+
+  public getCell({ key, columnIndex, rowIndex, style }) {
+    const data = this.getData(this.props.data);
+
+    const column = React.Children.toArray(this.props.children)[
+      columnIndex
+    ] as React.ReactElement<IColumnProps>;
+
+    if (!column || column.type !== Column) {
+      return null;
+    }
+
+    let className = css({
```

Comment:
Can we extract this into a function? I personally like the style a lot better with early returning and not overwriting and having that capsuled into a function as it makes it easier for me to quickly skim over the code here and still understand what it does

Reply:
could you please provide me with an example?

Reply:
Did that.